### ISSUE TITLE
Turn stylelint-config-standard into an actual standard

### ISSUE THREAD

**Issue by corysimmons** [stylelint/stylelint-config-standard#6]
> It favours flexibility over strictness for things like multi-line lists and single-line rulesets, and tries to avoid potentially divisive rules.

Is fine and all, but it's not a _standard_.

Flexible != Standard

You should rename this repo to `stylelint-config-flexible` and create an actual standard.

That's kind of the whole point behind linting anything to enforce strict rules for code quality.

Please take a page out of https://github.com/feross/standard handbook and become the one standard to rule them all.

**Comment by corysimmons** [comment]
That said, @dan-gamble has some good looking work on property ordering.

https://github.com/onespacemedia/project-template/blob/develop/%7B%7Bcookiecutter.repo_name%7D%7D/stylelint.config.js

I think combining what you've got already with some nice property grouping/ordering (alphabetical isn't useful) would be a good place to start.

**Comment by dan-gamble** [comment]
I'd be all for trying to enforce a standard like feross has done with JS (We use that where we work as we use Python and realise how useful PEP8 is)

Wouldn't mind contributing towards sorting out a CSStandard!

**Comment by MoOx** [comment]
imo the name is fine. The intention of ths config is good.
There is no real standard for CSS.
Even feross standard is an opininated config.
So maybe you should just discuss about enhancement instead of changing a name.
Or maybe, it can be renamed to something like "stylelint-config-recommended" like eslint:recommended.

**Comment by corysimmons** [comment]
> It favours flexibility over strictness for things like multi-line lists and single-line rulesets, and tries to avoid potentially divisive rules.

Okay, my enhancement is remove that line and add some property sorting by group and then alphabetical.

**Comment by corysimmons** [comment]
https://github.com/stylelint/stylelint-config-standard/issues/5#issuecomment-178496614 seems like it's not going to add sort rules/become a standard.

Oh well, I'm gonna get started on a standard. Thanks for the cool tool.

**Comment by MoOx** [comment]
For example feross standard don't force you to put all your var declarations at the same place.
You can extends this one and add you own rules.

**Comment by jeddy3** [comment]
@MoOx's That's a good point.

> #5 (comment) seems like it's not going to ... become a standard.

@corysimmons I'm not sure that's a fair assessment. This config provides a standard for 60 or so rules. It just avoids a divise rule, which I think is quite dependent on a team's choice of tooling e.g. quite a few PostCSS plugins use name-spaced custom declarations, should each of these be added to config?

Oh, I just noticed you built the lost grid. So, you'll understand the trouble with ordering an unknown number name-spaced custom declarations. Nice work with lost btw :)

**Comment by dan-gamble** [comment]
The guy who made `postcss-sorting` has a good way on how to handle `@at-rules` https://github.com/hudochenkov/postcss-sorting#at-rules

I'm guessing you could do the same with custom declarations as well. But it's a fair point the PostCSS ecosystem is huge and you'd have to cover a lot of things. I think Cory's point is valid in respect that property ordering doesn't really fall under the PostCSS ecosystem as it's a core part of CSS. But if you need to start ordering custom things that you can't really extend a config as you will be overwriting most of it

**Comment by jeddy3** [comment]
@dan-gamble Thanks for the input.

It feels like it's just the `rule-properties-order` rule that is contentious. Perhaps we can do something [similar to](https://github.com/feross/standard#i-want-to-contribute-to-standard-what-packages-should-i-know-about) `standard` and `semistandard`? i.e. you guys can create something like `stylelint-config-sortedstandard` which just adds declaration sorting to this config.

I can add something prominent to the README here pointing people in that direction. Handling the issue around sorting, and dealing with requests for custom things, can then be handled in the `stylelint-config-sortedstandard` repo.

I think there are good reasons for providing sorting and good reasons for not. At least this way we won't create entirely separate and competing "standards", and everyone can benefit from the same solid 60 rules.

What do you all think?

**Comment by dan-gamble** [comment]
I do think sorting does dive down more into team preferences. I don't think sorting in the global consensus will ever evolve out of preferring alphabetical with no groups.

I'd be up for trying to create a standard but i think rule ordering will never be a happy global decision. I think the standard ruleset for the most part is fine to linking off to things extending that would probably be the better way to go :+1:

**Comment by jeddy3** [comment]
> I think the standard ruleset for the most part is fine to linking off to things extending that would probably be the better way to go :+1: 

Good stuff :) @corysimmons Does this sound good to you?

**Comment by corysimmons** [comment]
I'm really not trying to be contrary here, but I really don't think giving people the choice to opt in to certain features makes it a standard. I mean... they can already do that.

Consider if browsers got to pick and choose which standards they implemented. Chrome might implement box-sizing nicely, Firefox might implement position: sticky nicely, but then Safari might decide to implement sub-pixel rounding their own way, and instantly we have a ton of headaches and StackOverflow questions.

That's only several browsers and we've had innumerable issues because they didn't agree on a standard. Hell, we coined it the "Browser Wars" and everyone was so ecstatic when browsers _finally_ started attempting to do things the same way.

That's only several browsers doing things slightly differently. With rule ordering, you'd have tens of thousands of developers doing things their own way. You'd never learn where to look for XYZ.

I genuinely believe this is something we can perfect over time. Most devs [prefer](https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/) to group by type. I know I do, and contrary to these issues, I really don't put a lot of thought into this stuff. I just do it naturally because it makes sense. I just wish there were rules about it.

Anyway, I closed these issues because I realized there's no point in even debating this. I want rules one way, and you want them another way. The entire purpose of creating configurable rules was to let anyone create their own way.

I'm just going to market my way as the official standard, and I hope that people will jump in the Issues to discuss ways we can improve it over time.

I'm pretty sure it'll be popular for the same reasons StandardJS is popular: people don't really care about what the standard is, as long as there is a pretty decent/popular/open-source one.

**Comment by corysimmons** [comment]
@jeddy3 What are your thoughts on adding `stylelint-config-strict`?

**Comment by jeddy3** [comment]
You make some good points :)

> I genuinely believe this is something we can perfect over time.

That is a great attitude to have and I'll do what I can to help you achieve it :)

The first hurdle to over come is the name as someone has already nabbed `stylelint-config-strict`. As you outlined above, "Standard" is probably the word you're after. Having "standard" will also allow you to piggy back on the traction of standardjs. And, as @MoOx , the word "recommended" is more aligned to the current goals of this repo and also parallels eslint's convention.

So, how best to go about this?...

This repo has one dependent already and a fair few downloads, but I think we can evolve the goal and therefore the strictness of the rules with a major version jump. Version `4.0.0` of standard will be more strict.

I can then create `stylelint-config-recommended` in the meantime to fulfil the current goals on this config i.e. "favouring flexibility over strictness for things like multi-line lists and single-line rulesets, and try to avoid potentially divisive rules."

Before I action anything though, I'd like to get @davidtheclark trusted opinion on this. What do you think David?

**Comment by davidtheclark** [comment]
Honestly, I do not care about the naming for this repo, between "standard", "recommended", something else like "base", "safe", etc. I just care that this repo continues doing what it's doing, because I think it's the most important extendable config: when there are 100 rules, providing a solid baseline that most people can extend without much or any overriding is very nice.

I think that a good migration pattern generally for stuff like this is to just to provide clear deprecation/renaming notices to users. It's the best we can do :)

I am fine with @corysimmons creating using the `stylelint-config-standard` name to create an opinionated config of his own. I am not ok with calling it "the official standard", though, because it's not. There is no "official standard" for stylelint.

(I myself would not use a config that enforces property ordering. I think that it's more trouble than its worth. So if *I* were making an opinionated standard, it would not include that. For what that's worth.)

Are we all on the same page then?

**Comment by corysimmons** [comment]
> I am not ok with calling it "the official standard", though, because it's not. There is no "official standard" for stylelint.

I see what you're saying. Personally I love that stylelint is so flexible (helps me integrate WP Coding Standards with my team). I never meant to insinuate that `stylelint-config-standard` should be the only ruleset we offer/maintain.

Instead of "this is the only ruleset" it should be "this is our strict ruleset".

So far as property ordering, a strict ruleset should definitely enforce that, and I'm in love with @dan-gamble 's property grouping/ordering/spacing. It'll all need tweaked to the right balance of strictness, but I really think the majority of users prefer grouping by type as CSS Trick's [poll](https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/) demonstrates.

![image](https://cloud.githubusercontent.com/assets/224636/12769700/e9080aac-c9e7-11e5-8ba1-5265c5544878.png)

I think the most interesting part is how unpopular alphabetical is.

Regardless, let's group by property, and sort alphabetically.

---

What's the next step forward from here? I'll get to work on that ruleset tonight and should have something pretty solid within a few hours.

**Comment by davidtheclark** [comment]
Cool! I think the next set is to sort out the configure you propose, write whatever docs you want; and meanwhile you and @jeddy3 can rename things as needed.

Thanks for working on this @corysimmons.

**Comment by corysimmons** [comment]
No problem!

I'll come up with a good ruleset and document rationale behind decisions.

Thank you all so much for being open minded about this. I really think this will be a great thing for the CSS community at large and stylelint in particular.

If we can get https://github.com/stylelint/stylelint/issues/330 (or https://github.com/morishitter/cssfmt/issues/61) sorted out as well I can easily see stylelint taking over the CSS linting scene.

:godmode:
```js
$ npm install --save-dev stylelint stylelint-config-standard

// package.json
"scripts" : {
  "styles:format": "stylelint --config standard css/**/*.css --scss --autocorrect"
}
```

**Comment by corysimmons** [comment]
Is it possible to disable the ability to overwrite rules with this particular ruleset?

**Comment by ntwb** [comment]
>  ...helps me integrate WP Coding Standards with my team...

@corysimmons you know that https://github.com/stylelint/stylelint-config-wordpress exists right?

**Comment by corysimmons** [comment]
@ntwb Yeah, that's what I was saying.

> Personally I love that stylelint is so flexible (helps me integrate WP Coding Standards with my team)

I love the fact that stylelint is flexible enough to write any config for it. In fact, it's helping my team adhere to the WP coding style.

**Comment by ntwb** [comment]
:+1:

**Comment by corysimmons** [comment]
hah... thought this would take a few hours tops. Parsing through hundreds of rules and coming up with a definitive standard is fun fun fun. :)

We should really rip off those CSSComb folks and make one of those config generators for the future http://csscomb.com/config :O

**Comment by ntwb** [comment]
> http://csscomb.com/config 
Thats cool :sunglasses:

**Comment by corysimmons** [comment]
Yeah, super user friendly. We'll get there!

**Comment by jeddy3** [comment]
It sounds like you've been busy and doing some great work Cory!

Up until now we haven't really given much thought to the role of the stylelint _org_ itself. We were too busy building stylelint :) But this feels like the right time to take stock and create a clear definition.

Much in the same way that `eslint` is the tool [that powers](https://github.com/feross/standard#i-want-to-contribute-to-standard-what-packages-should-i-know-about ) the likes of `standardjs`. We intend to create that same distinction with `stylelint` by moving all but the tool itself, the tester, the website, our _`eslint`_ config and an extendable baseline config out of the organisation.

Like I said above, I think your mission to create a config to unify the CSS community is a great one and it's awesome to know that stylelint is going to be used behind the scenes to power such a thing :)

And, like I also said above, I'll do what I can to help you get started. As such, the offer for you to use the `stylelint-config-standard` name still stands. You made some good points to why this name best represents your goals. And, as I pointed out, it'll let you piggy back on the traction of `standardjs`. 

Let me know if you think having the `stylelint-config-standard` name would be helpful to you and I'll move this repo over to your github user.

It'll be great if you could help us out as well by:

1. Using a major version bump and adding an explanation of the change in purpose to the changelog. I think this will be the best way of informing existing users of the config's change in purpose. It will also give them the opportunity to change to a extendable baseline, if that is what they need.
2. Making it clear within the documentation that is `stylelint-config-standard` isn't the "official standard" for stylelint. As David pointed out above, there is no "official standard" for stylelint :)

**Comment by dan-gamble** [comment]
I think the name would be a great help.

1. It's already an existing thing inside of the CSS linting ecosystem although the ruleset that follows with that could potentially dramatically change.
2. It aligns with JS name as well.

I wonder if it would be worth dropping the `-config` from it and just calling it `stylelint-standard` because  if we are looking to plough on with creating a standard i think that names has more meaning behind it. Also if everything is `stylelint-config-*` we could probably lose the excess word. :)

**Comment by MoOx** [comment]
having -config is a good idea imo. It avoid confusion. People can create plugin (eg: stylelint-plugin-whatever).
All smart too have some prefixes (eslint-config-, babel-preset- etc). We should not do the postcss-* mistake.

**Comment by jeddy3** [comment]
> I wonder if it would be worth dropping the -config from it and just calling it `stylelint-standard`

As @MoOx points out, the convention is to use `stylelint-config-*` for stylelint _configs_. However, much like `standardjs` [does](https://github.com/feross/standard#i-want-to-contribute-to-standard-what-packages-should-i-know-about), you're free to consume that config in another tool and name that tool as you wish.

I'd strongly recommend against naming that tool `stylelint-standard` though as that would give the wrong impresssion that it's `stylelint`'s standard. `stylelint` _does not have_ a standard.

Perhaps something like "standardcss" would be more apt when you reach the stage where you're consuming the config in another tool—perhaps once you have needs specific to your goal (for example, maybe something like [`standardjs`'s zero configuration](https://github.com/feross/standard#one-style-to-rule-them-all). 

In the beginning though, I think just providing a stylelint config called `stylelint-config-standard`, to match the config naming convention, will get you guys started :)

Does that make things a little clearer?

**Comment by dan-gamble** [comment]
It's a fair point and i misunderstood your last post.

> I'd strongly recommend against naming that tool stylelint-standard though as that would give the wrong impresssion that it's stylelint's standard. stylelint does not have a standard.

I thought you was getting behind Stylelint enforcing a standard, that clears it all up for me!

**Comment by jeddy3** [comment]
> i misunderstood your last post.

No worries, that happens a lot to me :) I'm trying to improve the clarity of my writing.

**Comment by dan-gamble** [comment]
I'm sure it was more on my behalf than yours, you always seem to be concise :+1:

**Comment by corysimmons** [comment]
@jeddy3 Little bummed `stylelint-config-standard` won't live here, but I'm happy to host it, and it'll free the org up to focus on stylelint as a tool rather than configs (you're migrating the other configs except the baseline out too right?).

Give me a few days (probably over this weekend) to finish up the first set of strict rules (going over 100+ rules turned out to be pretty intense, several hours in and halfway there) before we do the transfer.

Will definitely adhere to semver. Maybe you should throw an explanation up on the top of the README for a couple days for what's about to happen to the repo?

**Comment by jeddy3** [comment]
> you're migrating the other configs except the baseline out too right?).

Exactly. I'll be approaching new owners shortly e.g. it feels like the most logical home for the SuitCSS config is the [SuitCSS org]https://github.com/suitcss).

> it'll free the org up to focus on stylelint as a tool rather than configs

Yes, that's the intent here.

> Maybe you should throw an explanation up on the top of the README for a couple days 

Good idea. I'll draft something shortly.

**Comment by corysimmons** [comment]
:+1: :100:

**Comment by MoOx** [comment]
I still think that stylelint should offer a simple set of rules like eslint does with "eslint:recommended".

**Comment by jeddy3** [comment]
> I still think that stylelint should offer a simple set of rules like eslint does with "eslint:recommended".

Yep, that's exactly the intention :) 

As David [mentioned above](https://github.com/stylelint/stylelint-config-standard/issues/6#issuecomment-178903364), the stylelint org should continue to offer and maintain a recommended solid _baseline_ that most people can extend without much or any overriding. A baseline that "_favours flexibility over strictness for things like multi-line lists and single-line rulesets, and tries to avoid potentially divisive rules_".

I'm going to create `stylelint-config-recommended` for that purpose as it aligns with `eslint`.

**Comment by MoOx** [comment]
Good to heard.

**Comment by jeddy3** [comment]
@corysimmons Sorry to mess you around, but I've just realised that this name change is going to inconvenience more users than I originally thought. I hadn't noticed that it was downloaded a couple of thousand times in the last month when I offered to change the name and make `stylelint-config-standard` available to you. Sorry, I should of been a bit more thorough.

It sounds like you've taken the time to go over all the rules, which is great to here and I'm sure it'll help make your strict config very good! We've been working on stylelint for over a year and so the number of rules of rules has slowly grown over that period to a heady 100+ :)

I still think your mission to try and unify the CSS community is the good one, so I wish you the best of luck with your config and I hope you find a suitable and available name for it.

**Comment by MoOx** [comment]
One alternative solution:
- deprecate stylelint-config-standard in favor of stylelint-config-recommended on npm (and don't use stylelint-config-standard anymore)
- create stylelint-config-recommended for stylelint "official"/minimal recommended rules set
- create stylelint-config-standardcss as a "standardjs" like rules set

**Comment by jeddy3** [comment]
Thanks for alternative @MoOx. `stylelint-config-standardcss` is a good name for Cory's mission and perhaps he will take it :)

I'm in favour of not deprecating `stylelint-config-standard` though as it'll needlessly inconvenience those who already use it.

I think standard is fine as a name. I originally named it as standard as in _normal_ e.g. "You're wearing a standard looking suit today Geoff. Did you not want to put on something more fancy for a Friday?" :)

**Comment by MoOx** [comment]
I mentionned the deprecation to avoid the confusion that will be created by having 2 similars names stylelint-config-standard/stylelint-config-standardcss.

FYI, I did not have any issue/complain by deprecating cssnext package (with it downloaded like _a lot_) in favor of postcss-cssnext. Do not hesitate to do that. Deprecating a package does not break things. It warn the user for a change and the user can does it when he have time :)

**Comment by jeddy3** [comment]
> Deprecating a package does not break things. It warn the user for a change and the user can does it when he have time :)

Thanks, I didn't know that :)

**Comment by davidtheclark** [comment]
Agreed that deprecation is not a bad way to go. npm users should be accustomed to it.

**Comment by corysimmons** [comment]
I agree with @MoOx but I think I'm going to end up calling it `standard-css` instead of `stylelint-config-standardcss`.

I want people to be able to use it with any tool that supports the ruleset.

The whole `stylelint-config-standard` thing would still cause some confusion, and StandardCSS would first and foremost be written/maintained as a stylelint config, so it _would_ be nice if you could change the name of it with a deprecation warning.

I'll finish the config this weekend so you have a link to point people to.

**Comment by MoOx** [comment]
fyi, standard(js) have it's eslint config shared as eslint-config-standard.

**Comment by corysimmons** [comment]
@MoOx Ah thanks. Okay, so I'll practically mirror what feross is doing and create `standardcss` and `stylelint-config-standardcss` repos.

**Comment by dan-gamble** [comment]
:+1:

**Comment by corysimmons** [comment]
@jeddy3 You can keep the name. I'm done with stylelint. I'll check in every month or so to see if it's usable.

**Comment by MoOx** [comment]
That's not a very constructive feedback. All that "noise" to end with something which look like "stylelint is unusable, I am done with it" does not feel good. Can you elaborate?

**Comment by corysimmons** [comment]
I've thoroughly documented the problems I've had with stylelint in stylelint repo, Atom Linter repo, and Sublime Linter repo, as well as a couple other places.

The problem is you guys don't have official/functional linters for any of the major editors.

A linter without linters is not usable. How the hell did this slip by everyone?

I'm not upset that it's not usable. I'm happy you guys are working on a neat project to test out your JS/PostCSS chops. I'm upset that not one of you has actually used it and it's being marketed as a real world solution.

If I make a grid, I browser test it... I actually use it... I try to break it.. **then** I release it.. then I get feedback and work on it some more. Finally I make a website for it and market it. You guys made stylelint, skipped all that other stuff and went straight to marketing it.

**Comment by MoOx** [comment]
Wow.

> it's being marketed as a real world solution

Where did you see that?

> The problem is you guys don't have official/functional linters for any of the major editors.

Is this serious? Why don't you just do it? You are talking like "we" (note: I am not a maintainer) are supposed to provide everything "you" need.
People here have worked really hard and have developed an awesome tools with hundreds of rules like nobody made this before. Just look at other css linters... Oh there is no one that really works these days...

---

Most developers, actually use a terminal to build there project. That's where they see warnings and shit. 
And if you are more a "GUI" guy, you still have stuff like postcss-browser-reporter.

Saying that people code this and don't use this it just stupid. That's so obvious that "they" use it everyday.

Marketing is bullshit. There is no fucking marketing. They just share their works, that's it. The website is not even a real one (and nobody have said the opposite).

Sorry but you are talking like a dick that is looking to the "ultimate solution that fits all my needs".

There is no such things. Just contribute, like stylelint maintainers did, piece by piece, commit by commit, instead of saying "lol I am done with your unfinished untested work that no one can use".

**Comment by corysimmons** [comment]
No need to be sorry, I'm not a baby like the rest of GitHub, but if you wanna cuss me out in private I can get on gitter?

> "ultimate solution that fits all my needs"

Nah, I just want a linter that can lint in an editor rather than having to check a terminal task after every keystroke.

I would contribute if I were better at JS. I'm learning now, but time I could have spent studying was wasted trying to wrestle this project into basic functionality.

I was contributing in my own way though with a thought-out/strict standard that probably would've made stylelint a bit more popular.

**Comment by MoOx** [comment]
> I was contributing in my own way though with a thought-out/strict standard that probably would've made stylelint a bit more popular.

I said step by step, and here you are directly trying to achieve the final step. I agree that's a must have (I am very happy with eslint in atom) but you can't blame people for not doing something you want or need.

Btw, if you need a linter at every keystroke, maybe there is a problem with your CSS too ;)

**Comment by corysimmons** [comment]
> you can't blame people for not doing something you want or need.

For the 3rd time: I don't care that the tool isn't production ready. I care that there's no disclaimer warning others that style**lint** can't be used like a real linter since it doesn't work in any editor.

> if you need a linter at every keystroke, maybe there is a problem with your CSS too

I'm trying to introduce a real CSS standard with property sorting instead of these lax configs you guys are using.

**Comment by MoOx** [comment]
> I care that there's no disclaimer warning others that stylelint can't be used like a real linter since it doesn't work in any editor

Like I said, that's your fault. You have to check this before using a tool. For you information ESLint for example don't have an official atom linter plugin. The plugin is unofficial and had in the past lot's of failures. All those things demand time, and time is nor free or unlimited.

> I'm trying to introduce a real CSS standard with property sorting instead of these lax configs you guys are using.

Again, **step by step**.

**Comment by corysimmons** [comment]
I did try out the linters first. They seemed to work until you throw a custom config at them.

So in this order:

1. Found stylelint and assumed something called style**lint** could lint in a practical manner
2. Got plugins working for it
3. Spent days working on a stylelint config and reporting little bugs as I ran into them
4. Realized I had reported dozens of bugs and that I wouldn't be able to use my config for months

I'm not asking for an apology or anything. I just want you guys to put a disclaimer somewhere semi-prominent on your README and website that editor linting isn't supported and probably won't be for months.

**Comment by davidtheclark** [comment]
Ok, @corysimmons. Please go away.
